### ISSUE TITLE
Video freeze when WiFi signal is weak

### ISSUE THREAD

**Issue by dilbert168** [Genymobile/scrcpy#1730]
- [x] I have read the [FAQ](https://github.com/Genymobile/scrcpy/blob/master/FAQ.md).
 - [x] I have searched in existing [issues](https://github.com/Genymobile/scrcpy/issues).

**Environment**
 - OS: [Windows10]
 - scrcpy version: [1.16]
 - installation method: [Windows release]
 - device model:
 - Android version: [9]

**Describe the bug**
I have been using the scrcpy since V1.12.1 and recently update to V1.16.
What I have discovered the newer version behave less stable when WiFi signal is weak. (doing this for my project testing).  The video casting would freeze / pause.

When I enable the <ctrl-i> to check the fps info.  The v1.12.1 would drop the frame constantly when it is > 30 fps.
But in V1.16, it will not drop frame even signal is weak.  Is this new logic actually making it less stable ??
This is the only option I use " --bit-rate 8M"

Any option setting I could try out ??

**Comment by rom1v** [comment]
> What I have discovered the newer version behave less stable when WiFi signal is weak.

I guess it is not related to the difference between the version. It's just TCP end-of-line blocking on packet loss:
https://github.com/Genymobile/scrcpy/issues/495#issuecomment-482334870

> The v1.12.1 would drop the frame constantly when it is > 30 fps.

Dropping frame happens at rendering time (if you receive 3 frame before rendering one, it renders only the most recent, the two first frames are discarded).

> This is the only option I use " --bit-rate 8M"

8M is the default. If your wifi is not great, use a lower bitrate (`-b1M` for example) and also lower the definition (`-m 1024`), because a low bitrate on a high definition is worse.

**Comment by dilbert168** [comment]
I have tried to use the "--max-fps 25" to limit the frame transmitting.  However, my device is Android-9 and I understand this suppose to work only on Android-10.
On PC side, I enabled the <Alt-i> to see the fps and do see it limited ~25 from >30 fps before

My question is :  what could be expected behavior ?  Does device still sending >30 fps and only PC side not showing ?

Thank you

**Comment by dilbert168** [comment]
I just wondering, the server code running on Android, will it behave differently when connecting is USB-Cable and WiFi ?

**Comment by rom1v** [comment]
@dilbert168 No, it's not even aware that the connextion is over wifi or not.

**Comment by dilbert168** [comment]
Thanks for confirmation.
I would like to gather more debug info.  So I have add "-Vdebug"
In logcat, what filter I should put in to check for "scrcpy-server" possible ERROR or Warning during the issues.

**Comment by rom1v** [comment]
> I have tried to use the "--max-fps 25" to limit the frame transmitting. However, my device is Android-9 and I understand this suppose to work only on Android-10.

No it works on previous versions too (tested on Android 6).

> My question is : what could be expected behavior ? Does device still sending >30 fps and only PC side not showing ?

`--max-fps` limits the capture framerate (try `--max-fps 5` to see how it behaves).
